### PR TITLE
[llvm-amdgpu] Build device libs as external project

### PR DESCRIPTION
**Pull request by tpkessler**
* Build `rocm-device-libs` as an external subproject for LLVM. This way
  we put all files under `/opt/rocm/llvm/lib` where clang can find them.

* Split package. Build `rocm-device-libs` as an addtional package an
  symlink to `/opt/rocm/lib/cmake` where most other ROCm packages put
  their configuration files.

* Remove thread calculation for LLVM. This honors the user's configuration
  in makepkg.conf.
  CAUTION: Building LLVM takes a lot of RAM (40 GiB on 16 threads) and
  disc space (src directory has around 100 GiB) #197

### CODE REVIEW

**Comment by tpkessler** [comment]
I was able to build `comgr` the new `rocm-device-libs`

**Comment by acxz** [comment]
Should we also remove the standalone `rocm-device-libs` pkgbuild/.SRCINFO in this PR?

**Comment by tpkessler** [comment]
Yes, of course. And also from the AUR. Do you take care of this?

**Comment by tpkessler** [comment]
DO NOT MERGE
Installing `rocm-device-libs` in `/opt/rocm/llvm/lib` breaks `hip`. The package builds fine but `hipcc` cannot find the `*.bc` files. We need to add symlinks to `/opt/rocm/lib`

**Comment by acxz** [comment]
@tpkessler instead of adding symlinks everywhere can you do a thorough analysis with the installed files from the `llvm-amdgpu` deb package and our `llvm-amdgpu` package.

**Comment by ranisalt** [comment]
No! This will make an user needing device-libs only to have to build the entire LLVM project.

**Comment by tpkessler** [comment]
Hey @ranisalt we need `llvm-amdgpu` for `rocm-device-libs`, see [here](https://github.com/rocm-arch/rocm-arch/blob/9fed0486d2b8bfdd72864891b4197709a5125743/rocm-device-libs/PKGBUILD#L16)
So including the device libs into LLVM does not make a difference.

**Comment by tpkessler** [comment]
> can you do a thorough analysis with the installed files from the llvm-amdgpu deb package and our llvm-amdgpu package

Yes, I can do that. I've just downloaded the deb file.

**Comment by ranisalt** [comment]
You need `llvm-amdgpu` to *build* `rocm-device-libs`, not to *run* it. I want to be able to have `llvm-amdgpu` installed from a prebuilt package, and build `rocm-device-libs` on my own, without having to build the entire `llvm-amdgpu` again.

**Comment by tpkessler** [comment]
Are you sure that you don't need `llvm-amdgpu` to use the bitcode files `*.bc` from `rocm-device-libs`?

**Comment by tpkessler** [comment]
I don't know what I did yesterday. My changes on `llvm-amdgpu` were motivated by my work on `roctracer` and `rocprofiler`. But I've realized today that the patches I wrote are only needed to run some simple tests. The installed files are not affected by `rocm-device-libs`. I'll update my PRs for `roctracer` soon. I close this one now. Thank you for your reviews @acxz @ranisalt .

**Comment by ranisalt** [comment]
> Are you sure that you don't need `llvm-amdgpu` to use the bitcode files `*.bc` from `rocm-device-libs`?

I am pretty sure, unless something changed a lot. You can keep the changes, just don't need to merge the PKGBUILDs.

**Comment by tpkessler** [comment]
Which changes do you mean? All I did was adding `rocm-device-libs` to `llvm-amdgpu` (split package).

**Comment by ranisalt** [comment]
What about the external projects flag?

**Comment by tpkessler** [comment]
This flag is needed if you build `rocm-device-libs` as a LLVM subproject like `clang` or `lld`. This puts the `*.bc` under `/opt/rocm/llvm/libs` which is not in accordance which AMD's debian packages. Furthermore `hipcc` expects them to lie in `/opt/rocm/libs`. That's why I closed this PR.

**Comment by rigred** [comment]
I fixed the insane build requirements with 1 flag
Set `-DCMAKE_BUILD_TYPE=Release`

Otherwise it defaults to Debug
https://llvm.org/docs/GettingStarted.html#local-llvm-configuration
image

Please publish to AUR

**Comment by rigred** [comment]
We should verify and check that all projects here are compiled as their Release variants where possible.

**Comment by acxz** [comment]
> We should verify and check that all projects here are compiled as their Release variants where possible.

Regarding this see the following comment: https://github.com/rocm-arch/rocm-arch/issues/197#issuecomment-640574337

**Comment by rigred** [comment]
> TLDR by default Release should not be on, unless a special case can be proven such as has been done for llvm-amdgpu We will take requests for turning Release on case by case.

Agreed.
In the case of llvm, I've validated that Release is a valid option and carries no adverse consequences while providing a host of benefits.
The ubuntu llvm version of in rocm also appears to be a Release build.

We should anyhow not need our compiler built with Debug params.
It carries some ridiculous overhead.

**Comment by ranisalt** [comment]
Ideally, building a compiler is a process that should be left for beefy computers, loads of RAM and cores. The end user should use a prebuilt package. However, this is tricky to achieve using AUR. Arch4edu repo would help a lot.

For the LLVM package we can mimic whatever flags `llvm` from `extra` uses (and I believe it sets Release build type), that's a valid solution. But the rest of the packages should leave flags untouched.
